slsim/Sources/galaxies.py

```diff
@@ -120,14 +120,24 @@ def source_number_selected(self):
         """
         return self._num_select
 
-    def draw_source(self):
-        """Choose source at random.
+    def draw_source(self, z_max=None):
+        """Choose source at random. :param z_max: maximum redshift for source to be
+        drawn.
 
+        :param z_max: maximum redshift limit for the galaxy to be drawn. If no galaxy is
+          found for this limit, None will be returned.
         :return: dictionary of source
         """
-
-        index = random.randint(0, self._num_select - 1)
-        galaxy = self._galaxy_select[index]
+        if z_max is not None:
+            filtered_galaxies = self._galaxy_select[self._galaxy_select["z"] < z_max]
```

Comment:
for generating a large set of false positives, is this slowing the calculation down? You have to filter a potentially very long list first each time you draw a false positive. Just checking

Reply:
Yes, this slows down the calculation. z_max depands on each deflector that we draw. So, we need to do this here for each individual deflector. Currently, it takes 10 sec to draw 2000 false positive sample.